EIPS/eip-4788.md

```diff
@@ -77,76 +79,165 @@ When verifying a block, execution clients **MUST** ensure the root value in the
-sstore(HISTORY_STORAGE_ADDRESS, root_index, parent_beacon_block_root)
-```
+    timestamp_idx = to_uint256_be(timestamp) % HISTORICAL_ROOTS_MODULUS
+    timestamp = storage.get(timestamp_idx)
 
-#### New stateful precompile
+    if timestamp != evm.calldata:
+        evm.revert()
 
-Beginning at the execution timestamp `FORK_TIMESTAMP`, a "stateful" precompile is deployed at `HISTORY_STORAGE_ADDRESS`.
+    root_idx = timestamp_idx + HISTORICAL_ROOTS_MODULUS
+    root = storage.get(root_idx)
+    
+    evm.return(root)
 
-Callers of the precompile should provide the `timestamp` they are querying encoded as 32 bytes in big-endian format.
-Clients **MUST** sanitize this input call data to the precompile.
-If the input is _more_ than 32 bytes, the precompile only takes the first 32 bytes of the input buffer and ignores the rest.
-If the input is _less_ than 32 bytes, the precompile should revert.
+def set():
+    timestamp_idx = to_uint256_be(evm.timestamp) % HISTORICAL_ROOTS_MODULUS
+    root_idx = timestamp_idx + HISTORICAL_ROOTS_MODULUS
 
-Given this input, the precompile reduces the `timestamp` in the same way during the write routine and first checks if
-the `timestamp` recorded in the ring buffer matches the one supplied by the caller.
+    storage.set(timestamp_idx, evm.timestamp)
+    storage.set(root_idx, evm.calldata)
+```
 
-If the `timestamp` **does NOT** match, the client **MUST** return the "zero" word -- the 32-byte value where each byte is `0x00`.
+##### Bytecode
+
+The exact initcode to deploy is shared below.
+
+```asm
+push1 0x5a
+dup1
+push1 0x09
+push0
+codecopy
+push0
+return
+
+caller
+push20 0xfffffffffffffffffffffffffffffffffffffffe
+eq
+push1 0x42
+jumpi
+
+push1 0x20
+calldatasize
+eq
+push1 0x24
+jumpi
+
+push0
+push0
+revert
+
+jumpdest
+push3 0x018000
+push0
+calldataload
+mod
+dup1
+sload
+push0
+calldataload
+eq
+iszero
+push1 0x3d
+jumpi
+
+push3 0x018000
+add
+sload
+push0
+mstore
+
+jumpdest
+push1 0x20
+push0
+return
+
+jumpdest
+timestamp
+push3 0x018000
+timestamp
+mod
+sstore
+push0
+calldataload
+push3 0x018000
+timestamp
+mod
+push3 0x018000
+add
+sstore
+stop
+```
 
-If the `timestamp` **does** match, the client **MUST** read the root from the contract storage and return those 32 bytes in the caller's return buffer.
+#### Deployment
+
+The beacon roots contract is deployed like any other smart contract. A special synthetic address is generated
+by working backwards from the desired deployment transaction:
+
+```json
+{
+  "type": "0x2",
+  "chainId": "0x1",
+  "nonce": "0x0",
+  "to": null,
+  "gas": "0xd4f8",
+  "gasPrice": null,
+  "maxPriorityFeePerGas": "0x9c7652400",
+  "maxFeePerGas": "0xe8d4a51000",
+  "value": "0x0",
+  "input": "0x605a8060095f395ff33373fffffffffffffffffffffffffffffffffffffffe14604257602036146024575f5ffd5b620180005f350680545f351415603d576201800001545f525b60205ff35b42620180004206555f3562018000420662018000015500",
+  "accessList": [],
+  "v": "0x0",
+  "r": "0x539",
```

Comment:
> Although this style of contract creation is not tied to any specific initcode like create2 is, the synthetic address is cryptographically bound to the input data of the transaction (e.g. the initcode).

and then we don't have to think about things like this.

Reply:
>  rather than trying to use a TX of sorts but without gas or signature verification which requires a lot more exceptional logic

To be clear, this tx does still abide by gas and signature verification. The signature is constructed arbitrarily and therefore the sender is a one-time sender address (pk is not known).

--

In an effort to minimize protocol baggage, I think it is better to deploy the contract using standard methods. Once we are satisfied with the bytecode we'd like to deploy, it could be deployed by anyone. We simply have to agree on the address at which it is deployed. It doesn't have to be a synthetic tx, although I do slightly prefer this method as it is permissionless (anyone can fund the account and submit the tx).